### ISSUE TITLE
After update to v21.8.11 from v21.3.17 memory usage constantly increasing.

### ISSUE THREAD

**Issue by lexx-bright** [ClickHouse/ClickHouse#32750]
Mark_cache_size is set to default 5GB.
Dictionaries are not used.
The uncompressed cache is disabled.
Update to v21.8.12.29 didn't help.
![image](https://user-images.githubusercontent.com/59286375/146013519-01575276-fd92-44c0-ba03-0c2fecb045ba.png)
```
# ch_query () {
> clickhouse-client -m -n -q \
> "
> SELECT 'Allocated by dict' as alloc, formatReadableSize(sum(bytes_allocated)) FROM system.dictionaries;
>
> SELECT
>     database,
>     name,
>     formatReadableSize(total_bytes)
> FROM system.tables
> WHERE engine IN ('Memory','Set','Join');
>
> SELECT
>     'primary_key_bytes_in_memory' as pk_in_memory, formatReadableSize(sum(primary_key_bytes_in_memory)) AS primary_key_bytes_in_memory,
>     'primary_key_bytes_in_memory_allocated' as pk_allocated_in_memory, formatReadableSize(sum(primary_key_bytes_in_memory_allocated)) AS primary_key_bytes_in_memory_allocated
> FROM system.parts;
>
> SELECT
>     metric,
>     formatReadableSize(value)
> FROM system.asynchronous_metrics
> WHERE metric IN ('UncompressedCacheBytes', 'MarkCacheBytes');";
> }

# ch_query;\
> ps -q $(pidof clickhouse-server) -o rss=;\
> for i in {1..500};
>  do for j in {1..20};
>   do
>    (curl -m1 -s -F 'metrics_list=@mlist.tsv;' 'http://127.0.0.1:8123/?readonly=1&cancel_http_readonly_queries_on_client_close=1&metrics_list_structure=Path+String&query=SELECT+Path,+groupArray(Time),+groupArray(Value),+groupArray(Timestamp)+FROM+db.points+PREWHERE+Date+>=toDate(now()-interval+7+day)+AND+Date+<=+toDate(now())+WHERE+(Path+in+metrics_list)+AND+(Time+>=+toUnixTimestamp(now()-interval+7+day)+AND+Time+<=+toUnixTimestamp(now()))+GROUP+BY+Path+FORMAT+RowBinary' --output /tmp/query_result &>/dev/null & );
>   done;
>  sleep 1
>  if [[ $((${i}%25)) -eq 0 ]]; then
>   sleep 30; ps -q $(pidof clickhouse-server) -o rss=;
>  fi;
>  done;\
> sleep 60;\
> ch_query;\
> ps -q $(pidof clickhouse-server) -o rss=;

Allocated by dict       0.00 B
primary_key_bytes_in_memory     8.69 GiB        primary_key_bytes_in_memory_allocated   12.44 GiB
MarkCacheBytes  0.00 B
UncompressedCacheBytes  0.00 B
10891440
12313108
13180236
14087976
14972784
15837392
16788848
17643728
18468476
19435812
20124688
20968536
21812532
22665680
23506580
24395620
25245124
26062700
26964264
27797412
28368952
Allocated by dict       0.00 B
primary_key_bytes_in_memory     8.71 GiB        primary_key_bytes_in_memory_allocated   12.47 GiB
MarkCacheBytes  179.05 MiB
UncompressedCacheBytes  0.00 B
28365908
```

**Comment by den-crane** [comment]
>primary_key_bytes_in_memory_allocated   12.47 GiB

Check what table is allocated this memory 

```
SELECT database, table, sum(rows) rows, count() part_count, sum(marks) count_marks, formatReadableSize(sum(marks_bytes)) marks_size,
    formatReadableSize(sum(primary_key_bytes_in_memory)) AS bytes_in_memory,
    formatReadableSize((sum(primary_key_bytes_in_memory_allocated)) as s) AS bytes_in_memory_allocated
FROM system.parts
group by  database, table
order by s desc
limit 10;

```

**Comment by den-crane** [comment]
Also I don't see a result of 

```
 SELECT
     database,
     name,
     formatReadableSize(total_bytes)
 FROM system.tables
 WHERE engine IN ('Memory','Set','Join');
```

**Comment by lexx-bright** [comment]
@den-crane 
![image](https://user-images.githubusercontent.com/59286375/146038536-a0e239d1-cf90-4adf-ae29-555d6875ed91.png)
The query above is executed against db006.points_r1m_main table.
>Also I don't see a result of

![image](https://user-images.githubusercontent.com/59286375/146039099-4ebbd648-eeed-490f-93ee-5e4100aa02c1.png)

**Comment by den-crane** [comment]
Please share results as text. I need these numbers, I don't want to retype count_marks.

Also please share 
```
select metric, formatReadableSize(value) from system.asynchronous_metrics	where metric ilike '%mem%'
```

**Comment by lexx-bright** [comment]
@den-crane I've restarted a server and made a clean run.
```
Allocated by dict       0.00 B
primary_key_bytes_in_memory     8.34 GiB        primary_key_bytes_in_memory_allocated   11.89 GiB
MarkCacheBytes  0.00 B
UncompressedCacheBytes  0.00 B
10466216
11856436
12762412
13636148
14544048
15415700
16276760
17123264
17995268
18857280
19720840
20586020
21418904
22283652
23169352
24038052
24854208
25788364
26644836
27477828
28247468
Allocated by dict       0.00 B
primary_key_bytes_in_memory     8.35 GiB        primary_key_bytes_in_memory_allocated   11.90 GiB
MarkCacheBytes  179.05 MiB
UncompressedCacheBytes  0.00 B
28247980
```
```
SELECT
    database,
    name,
    formatReadableSize(total_bytes)
FROM system.tables
WHERE engine IN ('Memory', 'Set', 'Join')

Query id: cbf077da-fa68-44f8-b766-7eff7757efc1

Ok.

0 rows in set. Elapsed: 0.004 sec.
```
```
SELECT
    substring(database, 1, 5),
    table,
    sum(rows) AS rows,
    count() AS part_count,
    sum(marks) AS count_marks,
    formatReadableSize(sum(marks_bytes)) AS marks_size,
    formatReadableSize(sum(primary_key_bytes_in_memory)) AS bytes_in_memory,
    formatReadableSize(sum(primary_key_bytes_in_memory_allocated) AS s) AS bytes_in_memory_allocated
FROM system.parts
GROUP BY
    database,
    table
ORDER BY s DESC
LIMIT 10

Query id: 94179d1b-7d99-43eb-88ed-9e5b5b62eedf

┌─substring(database, 1, 5)─┬─table───────────┬────────rows─┬─part_count─┬─count_marks─┬─marks_size─┬─bytes_in_memory─┬─bytes_in_memory_allocated─┐
│ db006                     │ points_r1m_main │ 83630687029 │        101 │    10209107 │ 1.14 GiB   │ 2.06 GiB        │ 3.06 GiB                  │
│ db006                     │ points_r5m_max  │ 40788601967 │         70 │     4979215 │ 569.83 MiB │ 875.56 MiB      │ 1.17 GiB                  │
│ db006                     │ points_r5m_avg  │ 40879410594 │         82 │     4990306 │ 571.10 MiB │ 878.25 MiB      │ 1.11 GiB                  │
│ db006                     │ points_r15m_max │ 39414983748 │         84 │     4811535 │ 550.64 MiB │ 738.28 MiB      │ 1.03 GiB                  │
│ db006                     │ points_r15m_avg │ 39024403912 │         76 │     4763848 │ 545.18 MiB │ 729.79 MiB      │ 943.20 MiB                │
│ db006                     │ points_r1h_avg  │ 29126811328 │        114 │     3555697 │ 406.92 MiB │ 415.01 MiB      │ 575.08 MiB                │
│ db006                     │ points_r1h_max  │ 29115987533 │        111 │     3554364 │ 406.76 MiB │ 414.65 MiB      │ 569.83 MiB                │
│ db007                     │ points_r1m_main │ 19524246943 │         93 │     2383445 │ 272.75 MiB │ 201.83 MiB      │ 300.66 MiB                │
│ db002                     │ points_r1m_main │ 19636788916 │         88 │     2397190 │ 274.33 MiB │ 204.34 MiB      │ 296.17 MiB                │
│ db002                     │ points_r5m_max  │  9579527816 │         75 │     1169478 │ 133.83 MiB │ 99.80 MiB       │ 170.64 MiB                │
└───────────────────────────┴─────────────────┴─────────────┴────────────┴─────────────┴────────────┴─────────────────┴───────────────────────────┘
```
```
SELECT
    metric,
    formatReadableSize(value)
FROM system.asynchronous_metrics
WHERE metric ILIKE '%mem%'

Query id: 6363d747-852b-44bb-b339-53d5b0770d44

┌─metric────────────────────┬─formatReadableSize(value)─┐
│ MemoryResident            │ 26.93 GiB                 │
│ OSMemoryTotal             │ 31.25 GiB                 │
│ OSMemoryBuffers           │ 40.89 MiB                 │
│ OSMemoryFreePlusCached    │ 2.89 GiB                  │
│ OSMemorySwapCached        │ 13.89 MiB                 │
│ OSMemoryCached            │ 2.68 GiB                  │
│ OSMemoryAvailable         │ 2.59 GiB                  │
│ MemoryDataAndStack        │ 50.94 GiB                 │
│ MemoryShared              │ 208.25 MiB                │
│ MemoryCode                │ 343.84 MiB                │
│ MemoryVirtual             │ 51.74 GiB                 │
│ OSMemoryFreeWithoutCached │ 221.86 MiB                │
└───────────────────────────┴───────────────────────────┘
```

**Comment by lexx-bright** [comment]
@den-crane looks like ClickHouse not releasing memory properly when an external table is used in the query.
The same scenario executed with data placed in the query doesn't cause a memory leak.
```
# ch_query;\
> ps -q $(pidof clickhouse-server) -o rss=;\
> for i in {1..500};
>  do for j in {1..20};
>   do
>    (/usr/bin/curl -m1 -S -s --output /tmp/query_result --data-binary "@-" 'http://127.0.0.1:8123/?readonly=1&cancel_http_readonly_queries_on_client_close=1&max_query_size=10000000' <<< "SELECT Path, groupArray(Time), groupArray(Value), groupArray(Timestamp) FROM db006.points_r1m_main PREWHERE Date >= toDate(now()-interval 7 day) AND Date <= toDate(now()) WHERE (Path in (`sed "s/^\(.*\)$/'\1'/p" mlist.tsv | paste -d, -s`)) AND (Time >= toUnixTimestamp(now()-interval 7 day) AND Time <= toUnixTimestamp(now())) GROUP BY Path FORMAT RowBinary" &>/dev/null &)
>   done;
>  sleep 1
>  if [[ $((${i}%25)) -eq 0 ]]; then
>   sleep 30; ps -q $(pidof clickhouse-server) -o rss=;
>  fi;
>  done;\
> sleep 60;\
> ch_query;\
> ps -q $(pidof clickhouse-server) -o rss=;

Allocated by dict       0.00 B
primary_key_bytes_in_memory     8.68 GiB        primary_key_bytes_in_memory_allocated   12.28 GiB
MarkCacheBytes  0.00 B
UncompressedCacheBytes  0.00 B
10813828
11215320
11227316
11236816
11242076
11239592
11303388
11296316
11294752
11302040
11292824
11324096
11322748
11336752
11339572
11357716
11352016
11350920
11359612
11378964
11547160
Allocated by dict       0.00 B
primary_key_bytes_in_memory     8.84 GiB        primary_key_bytes_in_memory_allocated   12.56 GiB
MarkCacheBytes  165.09 MiB
UncompressedCacheBytes  0.00 B
11539364

SELECT
    metric,
    formatReadableSize(value)
FROM system.asynchronous_metrics
WHERE metric ILIKE '%mem%'

Query id: 168d1baf-0bc8-464b-8d39-a2cab8134f94

┌─metric────────────────────┬─formatReadableSize(value)─┐
│ MemoryResident            │ 10.75 GiB                 │
│ OSMemoryTotal             │ 31.25 GiB                 │
│ OSMemoryBuffers           │ 125.96 MiB                │
│ OSMemoryFreePlusCached    │ 18.72 GiB                 │
│ OSMemorySwapCached        │ 13.63 MiB                 │
│ OSMemoryCached            │ 17.61 GiB                 │
│ OSMemoryAvailable         │ 18.55 GiB                 │
│ MemoryDataAndStack        │ 29.86 GiB                 │
│ MemoryShared              │ 212.10 MiB                │
│ MemoryCode                │ 343.84 MiB                │
│ MemoryVirtual             │ 30.66 GiB                 │
│ OSMemoryFreeWithoutCached │ 1.11 GiB                  │
└───────────────────────────┴───────────────────────────┘
```

**Comment by den-crane** [comment]
> ClickHouse not releasing memory properly

This is expected. Jemalloc  https://github.com/jemalloc/jemalloc allocator does not release memory immediately. This behavior saves CPU by avoiding deallocation/allocation cycles.

>│ MemoryResident            │ 10.75 GiB                 │

10 GB is normal RAM usage for ClickHouse.


>primary_key_bytes_in_memory_allocated   12.28 GiB
>marks_size  1.14 GiB

All of this expected for tables with 83 630 687 029 ( 84 billions rows).

There is NO memory leak. Or you failed to demonstrate it.

**Comment by lexx-bright** [comment]
@den-crane 
10 GB when external table is not used in the query.
```
/usr/bin/curl -m1 -S -s --output /tmp/query_result --data-binary "@-" 'http://127.0.0.1:8123/?readonly=1&cancel_http_readonly_queries_on_client_close=1&max_query_size=10000000' <<< "SELECT Path, groupArray(Time), groupArray(Value), groupArray(Timestamp) FROM db006.points_r1m_main PREWHERE Date >= toDate(now()-interval 7 day) AND Date <= toDate(now()) WHERE (Path in (`sed "s/^\(.*\)$/'\1'/p" mlist.tsv | paste -d, -s`)) AND (Time >= toUnixTimestamp(now()-interval 7 day) AND Time <= toUnixTimestamp(now())) GROUP BY Path FORMAT RowBinary"

┌─metric────────────────────┬─formatReadableSize(value)─┐
│ MemoryResident            │ 10.75 GiB                 │
│ OSMemoryTotal             │ 31.25 GiB                 │
│ OSMemoryBuffers           │ 125.96 MiB                │
│ OSMemoryFreePlusCached    │ 18.72 GiB                 │
│ OSMemorySwapCached        │ 13.63 MiB                 │
│ OSMemoryCached            │ 17.61 GiB                 │
│ OSMemoryAvailable         │ 18.55 GiB                 │
│ MemoryDataAndStack        │ 29.86 GiB                 │
│ MemoryShared              │ 212.10 MiB                │
│ MemoryCode                │ 343.84 MiB                │
│ MemoryVirtual             │ 30.66 GiB                 │
│ OSMemoryFreeWithoutCached │ 1.11 GiB                  │
└───────────────────────────┴───────────────────────────┘
````

26 GB  when the same query is executed with the external table.
```
curl -m1 -s -F 'metrics_list=@mlist.tsv;' 'http://127.0.0.1:8123/?readonly=1&cancel_http_readonly_queries_on_client_close=1&metrics_list_structure=Path+String&query=SELECT+Path,+groupArray(Time),+groupArray(Value),+groupArray(Timestamp)+FROM+db.points+PREWHERE+Date+>=toDate(now()-interval+7+day)+AND+Date+<=+toDate(now())+WHERE+(Path+in+metrics_list)+AND+(Time+>=+toUnixTimestamp(now()-interval+7+day)+AND+Time+<=+toUnixTimestamp(now()))+GROUP+BY+Path+FORMAT+RowBinary' --output /tmp/query_result &>/dev/null

┌─metric────────────────────┬─formatReadableSize(value)─┐
│ MemoryResident            │ 26.93 GiB                 │
│ OSMemoryTotal             │ 31.25 GiB                 │
│ OSMemoryBuffers           │ 40.89 MiB                 │
│ OSMemoryFreePlusCached    │ 2.89 GiB                  │
│ OSMemorySwapCached        │ 13.89 MiB                 │
│ OSMemoryCached            │ 2.68 GiB                  │
│ OSMemoryAvailable         │ 2.59 GiB                  │
│ MemoryDataAndStack        │ 50.94 GiB                 │
│ MemoryShared              │ 208.25 MiB                │
│ MemoryCode                │ 343.84 MiB                │
│ MemoryVirtual             │ 51.74 GiB                 │
│ OSMemoryFreeWithoutCached │ 221.86 MiB                │
└───────────────────────────┴───────────────────────────┘
```

How would you explain the difference?

**Comment by lexx-bright** [comment]
10000 queries run 
|With external data | Without external data|
|----------|-----------|
|Allocated by dict       0.00 B                                                                              |Allocated by dict       0.00 B                                                                   |
|primary_key_bytes_in_memory     8.34 GiB        primary_key_bytes_in_memory_allocated   11.89 GiB           |primary_key_bytes_in_memory     8.68 GiB        primary_key_bytes_in_memory_allocated   12.28 GiB|
|MarkCacheBytes  0.00 B                                                                                      |MarkCacheBytes  0.00 B                                                                           |
|UncompressedCacheBytes  0.00 B                                                                              |UncompressedCacheBytes  0.00 B                                                                   |
|10466216                                                                                                    |10813828                                                                                         |
|11856436                                                                                                    |11215320                                                                                         |
|12762412                                                                                                    |11227316                                                                                         |
|13636148                                                                                                    |11236816                                                                                         |
|14544048                                                                                                    |11242076                                                                                         |
|15415700                                                                                                    |11239592                                                                                         |
|16276760                                                                                                    |11303388                                                                                         |
|17123264                                                                                                    |11296316                                                                                         |
|17995268                                                                                                    |11294752                                                                                         |
|18857280                                                                                                    |11302040                                                                                         |
|19720840                                                                                                    |11292824                                                                                         |
|20586020                                                                                                    |11324096                                                                                         |
|21418904                                                                                                    |11322748                                                                                         |
|22283652                                                                                                    |11336752                                                                                         |
|23169352                                                                                                    |11339572                                                                                         |
|24038052                                                                                                    |11357716                                                                                         |
|24854208                                                                                                    |11352016                                                                                         |
|25788364                                                                                                    |11350920                                                                                         |
|26644836                                                                                                    |11359612                                                                                         |
|27477828                                                                                                    |11378964                                                                                         |
|28247468                                                                                                    |11547160                                                                                         |
|Allocated by dict       0.00 B                                                                              |Allocated by dict       0.00 B                                                                   |
|primary_key_bytes_in_memory     8.35 GiB        primary_key_bytes_in_memory_allocated   11.90 GiB           |primary_key_bytes_in_memory     8.84 GiB        primary_key_bytes_in_memory_allocated   12.56 GiB|
|MarkCacheBytes  179.05 MiB                                                                                  |MarkCacheBytes  165.09 MiB                                                                       |
|UncompressedCacheBytes  0.00 B                                                                              |UncompressedCacheBytes  0.00 B                                                                   |
|28247980                                                                                                    |11539364                                                                                         |

**Comment by den-crane** [comment]
I see. How big is mlist.tsv ?

```
ls -la mlist.tsv
wc -l mlist.tsv 
```

**Comment by den-crane** [comment]
It seems it reproduce with external data + readonly

```
 create table t (Path String) Engine=MergeTree order by Path;
 insert into t select toString(number) from numbers(1000000);

 for i in `seq 10000`; do echo $i;  \ 
    clickhouse-client -q "select formatReadableSize(value) FROM system.asynchronous_metrics WHERE metric='MemoryResident'"; \ 
    curl -m0.5 -s -F 's=@t.tsv;' 'http://127.0.0.1:8123/?readonly=1&s_structure=Path+String&query=SELECT+count()+FROM+t+WHERE+(Path+in+s)' -o /dev/null; \
done

**Comment by den-crane** [comment]
my test was incorrect

**Comment by den-crane** [comment]
```
create table t (Path String) Engine=Null;
clickhouse-client -q "select toString(number) from numbers(10000000) format TSV" > t.tsv

for i in `seq 10000`; do echo $i;   clickhouse-client -q "select formatReadableSize(value) FROM system.asynchronous_metrics WHERE metric='MemoryResident'";  curl  -s -F 's=@t.tsv;' 'http://127.0.0.1:8123/?cancel_http_readonly_queries_on_client_close=1&readonly=1&s_structure=Path+String&query=SELECT+count()+FROM+t+WHERE+(Path+in+s)' -o /dev/null; done
1
298.20 MiB
2
1.11 GiB
3
1.29 GiB
4
1.37 GiB
5
1.69 GiB
6
......
88
14.81 GiB
89
14.74 GiB
....
16.12 GiB
100
15.96 GiB
101
16.58 GiB
102
16.72 GiB
103
17.17 GiB
```

**Comment by den-crane** [comment]
I cannot reproduce with 21.11.5.33.
@lexx-bright check with 21.11.5.33

**Comment by den-crane** [comment]
@azat

**Comment by den-crane** [comment]
seems related to cancel_http_readonly_queries_on_client_close

**Comment by den-crane** [comment]
the same with 21.11.5.33

```
clickhouse-client -q "create table t (Path String) Engine=Null"
clickhouse-client -q "select toString(number) from numbers(10000000) format TSV" > t.tsv

for i in `seq 10000`; do echo $i;   clickhouse-client -q "select formatReadableSize(value) FROM system.asynchronous_metrics WHERE metric='MemoryResident'";  curl  -s -F 's=@t.tsv;' 'http://127.0.0.1:8123/?cancel_http_readonly_queries_on_client_close=1&readonly=1&s_structure=Path+String&query=SELECT+count()+FROM+t+WHERE+(Path+in+s)' -o /dev/null; done


1
344.61 MiB
2
344.61 MiB
3
344.61 MiB
4
344.61 MiB
5
344.61 MiB
6
....
27.38 GiB
120
27.38 GiB
121
27.38 GiB
122
27.38 GiB
123
27.38 GiB
124
27.38 GiB
125
27.38 GiB
126
27.38 GiB
127
27.38 GiB
128
27.38 GiB
```

**Comment by azat** [comment]
@den-crane thanks for the test, will take a look.

**Comment by PhantomPhreak** [comment]
This bug has been reported in v21.8.11, recent release is v21.8.15.7, but this bugfix wasn't backported yet.
https://github.com/ClickHouse/ClickHouse/compare/v21.8.11.4-lts...v21.8.15.7-lts

Recent v21.8.15.7 still leaking memory with `cancel_http_readonly_queries_on_client_close` option enabled

21.8 branch is LTS now (LTS was recommended to use by ClickHouse developers), but we can't update to 21.8-lts from 21.3-lts because of this bug. Is there any plans to add this fix to the 21.8 branch?
Thanks!
